dialogue-target/src/main/java/com/palantir/dialogue/Request.java

```diff
@@ -249,81 +243,81 @@ private Builder() {}
         }
 
         private ListMultimap<String, String> mutableHeaderParams() {
-            if (headerParams == null) {
-                headerParams = MultimapBuilder.treeKeys(String.CASE_INSENSITIVE_ORDER)
-                        .arrayListValues()
-                        .build();
-
-                if (existingUnmodifiableHeaderParams != null) {
-                    existingUnmodifiableHeaderParams.forEach(headerParams::put);
-                    existingUnmodifiableHeaderParams = null;
+            if (!isHeaderMutable()) {
+                setHeaderMutable();
+                ListMultimap<String, String> mutable =
+                        Multimaps.newListMultimap(new TreeMap<>(String.CASE_INSENSITIVE_ORDER), MAP_VALUE_FACTORY);
+                if (!headerParams.isEmpty()) {
+                    // Outperforms mutable.putAll(headerParams)
+                    headerParams.forEach(mutable::put);
                 }
+                headerParams = mutable;
             }
             return headerParams;
         }
 
-        private ImmutableListMultimap.Builder<String, String> mutableQueryParams() {
-            if (queryParams == null) {
-                queryParams = ImmutableListMultimap.builder();
-
-                if (existingQueryParams != null) {
-                    existingQueryParams.forEach(queryParams::put);
-                    existingQueryParams = null;
-                }
+        private ListMultimap<String, String> mutableQueryParams() {
+            if (!isQueryMutable()) {
+                setQueryMutable();
+                queryParams = ArrayListMultimap.create(queryParams);
             }
             return queryParams;
         }
 
-        private ImmutableMap.Builder<String, String> mutablePathParams() {
-            if (pathParams == null) {
-                pathParams = ImmutableMap.builder();
-
-                if (existingPathParams != null) {
-                    existingPathParams.forEach(pathParams::put);
-                    existingPathParams = null;
-                }
+        private Map<String, String> mutablePathParams() {
+            if (!isPathMutable()) {
+                setPathMutable();
+                pathParams = new HashMap<>(pathParams);
```

Comment:
I believe ImmutableMap preserves insertion order... should this be a LinkedHashMap just in case?

Reply:
If order is important for any of these collections we have some bigger problems!